examples/vision/retinanet.py

```diff
@@ -0,0 +1,700 @@
+        matched_gt_idx = tf.argmax(iou_matrix, axis=1)
+        positive_mask = tf.greater_equal(max_iou, match_iou)
+        negative_mask = tf.less(max_iou, ignore_iou)
+        ignore_mask = tf.logical_not(tf.logical_or(positive_mask, negative_mask))
+        return (
+            matched_gt_idx,
+            tf.cast(positive_mask, dtype=tf.float32),
+            tf.cast(ignore_mask, dtype=tf.float32),
+        )
+
+    def _compute_box_target(self, anchor_boxes, matched_gt_boxes):
+        box_target = tf.concat(
+            [
+                (matched_gt_boxes[:, :2] - anchor_boxes[:, :2]) / anchor_boxes[:, 2:],
+                tf.math.log(matched_gt_boxes[:, 2:] / anchor_boxes[:, 2:]),
+            ],
+            axis=-1,
+        )
+        box_target = box_target / self._box_variance
+        return box_target
+
+    def _encode_sample(self, image_shape, gt_boxes, cls_ids):
+        """Creates box and classification targets for a single sample"""
+        h = image_shape[1]
+        w = image_shape[2]
+        anchor_boxes = self._anchor_box.get_anchors(h, w)
+        cls_ids = tf.cast(cls_ids, dtype=tf.float32)
+        matched_gt_idx, positive_mask, ignore_mask = self._match_anchor_boxes(
+            anchor_boxes, gt_boxes
+        )
+        matched_gt_boxes = tf.gather(gt_boxes, matched_gt_idx)
+        box_target = self._compute_box_target(anchor_boxes, matched_gt_boxes)
+        matched_gt_cls_ids = tf.gather(cls_ids, matched_gt_idx)
+        cls_target = tf.where(
+            tf.not_equal(positive_mask, 1.0), -1.0, matched_gt_cls_ids
+        )
+        cls_target = tf.where(tf.equal(ignore_mask, 1.0), -2.0, cls_target)
+        cls_target = tf.expand_dims(cls_target, axis=-1)
+        label = tf.concat([box_target, cls_target], axis=-1)
+        return label
+
+    def encode_batch(self, batch_images, gt_boxes, cls_ids):
+        """Creates box and classification targets for a batch"""
+        images_shape = tf.shape(batch_images)
+        batch_size = images_shape[0]
+
+        labels = tf.TensorArray(dtype=tf.float32, size=batch_size, dynamic_size=True)
+        for i in range(batch_size):
+            label = self._encode_sample(images_shape, gt_boxes[i], cls_ids[i])
+            labels = labels.write(i, label)
+
+        batch_images = tf.keras.applications.resnet.preprocess_input(batch_images)
+        return batch_images, labels.stack()
+
+
+"""
+## Building the ResNet50 backbone
+"""
+
+
+def get_backbone():
+    """Builds ResNet50 with pre-trained imagenet weights"""
+    backbone = keras.applications.ResNet50(
+        include_top=False, input_shape=[None, None, 3]
+    )
+    C3, C4, C5 = [
+        backbone.get_layer(layer_name).output
+        for layer_name in ["conv3_block4_out", "conv4_block6_out", "conv5_block3_out"]
+    ]
+    return keras.Model(inputs=[backbone.inputs], outputs=[C3, C4, C5])
+
+
+"""
+## Building Feature Pyramid Network as a custom layer
+"""
+
+
+class FPN(keras.layers.Layer):
```

Comment:
Prefer using a specific naming, i.e., ResNet50FeaturePyramid, because it is highly model specific.